### PR TITLE
Limit JSON RPC message size

### PR DESCRIPTION
**Pull request by N-o-X**

### CODE REVIEW

File: lib/base/netstring.cpp [hunk 1]
```diff
@@ -84,6 +85,15 @@ StreamReadStatus NetString::ReadStringFromStream(const Stream::Ptr& stream, Stri
 	/* read the whole message */
 	size_t data_length = len + 1;
 
+	if (data_length > maxMessageLength) {
+		int maxLengthInMegaByte = maxMessageLength / 1024;
```
Comment (gunnarbeutner) [inline lib/base/netstring.cpp]: That'd be kB, wouldn't it?

File: lib/remote/jsonrpcconnection.cpp [hunk 2]
```diff
@@ -250,9 +250,14 @@ void JsonRpcConnection::MessageHandler(const String& jsonString)
 
 bool JsonRpcConnection::ProcessMessage()
 {
+	size_t maxMessageLength = 64 * 1024;
+
+	if (m_Endpoint)
+		maxMessageLength = -1;
```
Comment (gunnarbeutner) [inline lib/remote/jsonrpcconnection.cpp]: We really shouldn't rely on the overflow behavior of signed-to-unsigned conversion. Use SIZE_MAX instead.

**Comment by dnsmichi** [comment]
Please also add some test result of yours, especially the log messages when a too large chunk is received.

**Comment by Crunsher** [comment]
Using a user with query permissions only and a very big file for upload I observed the following behavior. The only difference in the calls was changing the Content Length Header.
```
curl command used:
curl -k -H "Content-length: $length" -v -u query:icinga -H 'Accept: application/json' -X POST 'https://localhost:5665/v1/objects/hosts' -d @fil -i

length = 0:
{"error":404.0,"status":"No objects found."}

length = -1:
<h1>Content length exceeded maximum</h1>

length = 99999999999999999999:
<h1>Bad Request</h1><p><pre>Content-Length must not be negative.</pre></p>

length = 5:
json error at $5th char

length = 1025
json error in the middle of the first line

length = 2015
json error in the middle of the first line

length = 1048576 (1024^2, 1MiB max length for this user)
<h1>Bad Request</h1><p><pre>Line length for HTTP header exceeded</pre></p>

length = 1048577
<h1>Content length exceeded maximum</h1>
```

**Comment by Crunsher** [comment]
Ugly stacktrace in log:
```
[2018-03-05 13:10:21 +0100] warning/JsonRpcConnection: Error while reading JSON-RPC message for identity '': Error: Max data length exceeded: 0 MB

	(0) icinga2: void boost::throw_exception<boost::exception_detail::error_info_injector<std::invalid_argument> >(boost::exception_detail::error_info_injector<std::invalid_argument> const&) (+0x6c) [0x561561fcdc23]
	(1) icinga2: void boost::exception_detail::throw_exception_<std::invalid_argument>(std::invalid_argument const&, char const*, char const*, int) (+0xb0) [0x561561fcc6b3]
	(2) icinga2: icinga::NetString::ReadStringFromStream(boost::intrusive_ptr<icinga::Stream> const&, icinga::String*, icinga::StreamReadContext&, bool, long) (+0x3b4) [0x56156209c348]
	(3) icinga2: icinga::JsonRpc::ReadMessage(boost::intrusive_ptr<icinga::Stream> const&, icinga::String*, icinga::StreamReadContext&, bool, unsigned long) (+0x49) [0x5615622912cb]
	(4) icinga2: icinga::JsonRpcConnection::ProcessMessage() (+0xa7) [0x5615622936f1]
	(5) icinga2: icinga::JsonRpcConnection::DataAvailableHandler() (+0x83) [0x561562293913]
	(6) icinga2: icinga::JsonRpcConnection::Start() (+0xd0) [0x561562291db0]
	(7) icinga2: icinga::ApiListener::NewClientHandlerInternal(boost::intrusive_ptr<icinga::Socket> const&, icinga::String const&, icinga::ConnectionRole) (+0xace) [0x5615621f2604]
	(8) icinga2: icinga::ApiListener::NewClientHandler(boost::intrusive_ptr<icinga::Socket> const&, icinga::String const&, icinga::ConnectionRole) (+0x4a) [0x5615621f1938]
	(9) icinga2: void std::__invoke_impl<void, void (icinga::ApiListener::*&)(boost::intrusive_ptr<icinga::Socket> const&, icinga::String const&, icinga::ConnectionRole), icinga::ApiListener*&, boost::intrusive_ptr<icinga::Socket>&, icinga::String&, icinga::ConnectionRole&>(std::__invoke_memfun_deref, void (icinga::ApiListener::*&)(boost::intrusive_ptr<icinga::Socket> const&, icinga::String const&, icinga::ConnectionRole), icinga::ApiListener*&, boost::intrusive_ptr<icinga::Socket>&, icinga::String&, icinga::ConnectionRole&) (+0xad) [0x56156221829d]
	(10) icinga2: std::__invoke_result<void (icinga::ApiListener::*&)(boost::intrusive_ptr<icinga::Socket> const&, icinga::String const&, icinga::ConnectionRole), icinga::ApiListener*&, boost::intrusive_ptr<icinga::Socket>&, icinga::String&, icinga::ConnectionRole&>::type std::__invoke<void (icinga::ApiListener::*&)(boost::intrusive_ptr<icinga::Socket> const&, icinga::String const&, icinga::ConnectionRole), icinga::ApiListener*&, boost::intrusive_ptr<icinga::Socket>&, icinga::String&, icinga::ConnectionRole&>(void (icinga::ApiListener::*&)(boost::intrusive_ptr<icinga::Socket> const&, icinga::String const&, icinga::ConnectionRole), icinga::ApiListener*&, boost::intrusive_ptr<icinga::Socket>&, icinga::String&, icinga::ConnectionRole&) (+0x87) [0x561562214f25]
	(11) icinga2: void std::_Bind<void (icinga::ApiListener::*(icinga::ApiListener*, boost::intrusive_ptr<icinga::Socket>, icinga::String, icinga::ConnectionRole))(boost::intrusive_ptr<icinga::Socket> const&, icinga::String const&, icinga::ConnectionRole)>::__call<void, , 0ul, 1ul, 2ul, 3ul>(std::tuple<>&&, std::_Index_tuple<0ul, 1ul, 2ul, 3ul>) (+0xce) [0x5615622103a4]
	(12) icinga2: void std::_Bind<void (icinga::ApiListener::*(icinga::ApiListener*, boost::intrusive_ptr<icinga::Socket>, icinga::String, icinga::ConnectionRole))(boost::intrusive_ptr<icinga::Socket> const&, icinga::String const&, icinga::ConnectionRole)>::operator()<, void>() (+0x2a) [0x56156220af58]
	(13) icinga2: void std::__invoke_impl<void, std::_Bind<void (icinga::ApiListener::*(icinga::ApiListener*, boost::intrusive_ptr<icinga::Socket>, icinga::String, icinga::ConnectionRole))(boost::intrusive_ptr<icinga::Socket> const&, icinga::String const&, icinga::ConnectionRole)>>(std::__invoke_other, std::_Bind<void (icinga::ApiListener::*(icinga::ApiListener*, boost::intrusive_ptr<icinga::Socket>, icinga::String, icinga::ConnectionRole))(boost::intrusive_ptr<icinga::Socket> const&, icinga::String const&, icinga::ConnectionRole)>&&) (+0x20) [0x561562206650]
	(14) icinga2: std::__invoke_result<std::_Bind<void (icinga::ApiListener::*(icinga::ApiListener*, boost::intrusive_ptr<icinga::Socket>, icinga::String, icinga::ConnectionRole))(boost::intrusive_ptr<icinga::Socket> const&, icinga::String const&, icinga::ConnectionRole)>>::type std::__invoke<std::_Bind<void (icinga::ApiListener::*(icinga::ApiListener*, boost::intrusive_ptr<icinga::Socket>, icinga::String, icinga::ConnectionRole))(boost::intrusive_ptr<icinga::Socket> const&, icinga::String const&, icinga::ConnectionRole)>>(std::_Bind<void (icinga::ApiListener::*(icinga::ApiListener*, boost::intrusive_ptr<icinga::Socket>, icinga::String, icinga::ConnectionRole))(boost::intrusive_ptr<icinga::Socket> const&, icinga::String const&, icinga::ConnectionRole)>&&) (+0x26) [0x561562202657]
	(15) icinga2: decltype (__invoke((_S_declval<0ul>)())) std::thread::_Invoker<std::tuple<std::_Bind<void (icinga::ApiListener::*(icinga::ApiListener*, boost::intrusive_ptr<icinga::Socket>, icinga::String, icinga::ConnectionRole))(boost::intrusive_ptr<icinga::Socket> const&, icinga::String const&, icinga::ConnectionRole)> > >::_M_invoke<0ul>(std::_Index_tuple<0ul>) (+0x28) [0x561562220774]
	(16) icinga2: std::thread::_Invoker<std::tuple<std::_Bind<void (icinga::ApiListener::*(icinga::ApiListener*, boost::intrusive_ptr<icinga::Socket>, icinga::String, icinga::ConnectionRole))(boost::intrusive_ptr<icinga::Socket> const&, icinga::String const&, icinga::ConnectionRole)> > >::operator()() (+0x1d) [0x561562220493]
	(17) icinga2: std::thread::_State_impl<std::thread::_Invoker<std::tuple<std::_Bind<void (icinga::ApiListener::*(icinga::ApiListener*, boost::intrusive_ptr<icinga::Socket>, icinga::String, icinga::ConnectionRole))(boost::intrusive_ptr<icinga::Socket> const&, icinga::String const&, icinga::ConnectionRole)> > > >::_M_run() (+0x1c) [0x5615622201b6]
	(18) libstdc++.so.6: <unknown function> (+0xbc773) [0x7fa57a5e5773]
	(19) libpthread.so.0: <unknown function> (+0x751a) [0x7fa57b69951a]
	(20) libc.so.6: clone (+0x3f) [0x7fa579d053ef]

```

**Pull request merged**